### PR TITLE
[1] docs: files introduction

### PR DESCRIPTION
**Pull request by topless**
- [x] Integrating files into your application
- [x] Serving files (to take from the modules docs and rework)
- [x] Configuring your storage backend (and multiple backends)

addresses #3972

### CODE REVIEW

File: docs/files/backends.rst [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    This file is part of Invenio.
+    Copyright (C) 2019 CERN.
+
+    Invenio is free software; you can redistribute it and/or modify it
+    under the terms of the MIT License; see LICENSE file for more details.
+
+.. _backends:
+
+
+Configure your Storage Backend
```
Comment (ChiaraBi) [inline docs/files/backends.rst]: It's not clear for me if this section was meant to be a "how-to guide" or a "reference guide" (I am referring to the classification in the article https://www.divio.com/blog/documentation/ ). We can discuss about it :)

File: docs/files/backends.rst [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+==============================
+
+A storage backend serves as an interface for the underlying file access.
+
+In the `config.py` file of your application, you can use the variable
```
Comment (equadon) [inline docs/files/backends.rst]:
```suggestion
In the ``config.py`` file of your application, you can use the variable
```

File: docs/files/backends.rst [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+`FILES_REST_STORAGE_FACTORY <https://invenio-files-rest.readthedocs.io/en/latest/configuration.html#invenio_files_rest.config.FILES_REST_STORAGE_FACTORY>`_
+to provide your custom factory that will be used to create the storage
+instance.
+
+The module provides an abstract layer for storage implementation that allows
```
Comment (equadon) [inline docs/files/backends.rst]:
```suggestion
The module provides an abstraction layer for storage implementation that allows
```
?

File: docs/files/backends.rst [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+for Invenio-Files-REST is `PyFilesystem <https://www.pyfilesystem.org/>`_,
+which can be replaced according to your needs even with a cloud
+storage like `Invenio-S3 <https://invenio-s3.readthedocs.io/>`_ which
+offers integration with any S3 REST API compatible object storage.
+
+
+Build your own Storage Backend
+------------------------------
+
+If you would like to provide your own implementation as a storage backend, it
```
Comment (equadon) [inline docs/files/backends.rst]:
```suggestion
If you would like to provide your own implementation of a storage backend, it
```

File: docs/files/backends.rst [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+is requires to subclass the `FileStorage <https://invenio-files-rest.readthedocs.io/en/latest/api.html#invenio_files_rest.ext.FileStorage>`_
```
Comment (equadon) [inline docs/files/backends.rst]:
```suggestion
is required to subclass the `FileStorage <https://invenio-files-rest.readthedocs.io/en/latest/api.html#invenio_files_rest.ext.FileStorage>`_
```

File: docs/files/backends.rst [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+class, and implement for some of its core methods.
```
Comment (equadon) [inline docs/files/backends.rst]:
```suggestion
class, and implement some of its core methods.
```

File: docs/files/backends.rst [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+The following methods are required to be implemented for the storage to work:
+
+* :code:`initialize`
+* :code:`open`
+* :code:`save`
+* :code:`update`
+* :code:`delete`
+
+The base class provides also some optional methods to implement to extend the
```
Comment (equadon) [inline docs/files/backends.rst]:
```suggestion
The base class provides also some optional methods to extend the
```

File: docs/files/index.rst [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    Copyright (C) 2019 CERN.
+
+    Invenio is free software; you can redistribute it and/or modify it
+    under the terms of the MIT License; see LICENSE file for more details.
+
+Handling Files
+==============
+
+.. toctree::
+   :maxdepth: 1
```
Comment (equadon) [inline docs/files/index.rst]: What do you think about having `maxdepth` set to `2` instead so we can see the sub-sections and more easily navigate?

File: docs/files/integration.rst [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+Integrating Files
+=================
+
+Invenio provides a bundle of modules to accommodate all around needs about
+file management. You can install all the necessary modules at once, or due to
+invenio modular architecture you can hand pick only those you need.
+
+In case you have scaffolded your invenio application with
+`cookiecutter-invenio-instance <https://github.com/inveniosoftware/cookiecutter-invenio-instance>`_
+files bundle is already a part of your solution, the only step you need to
```
Comment (equadon) [inline docs/files/integration.rst]: This assumes that the reader is starting from cookiecutter with Invenio 3.2. If they created their project with Invenio 3.1 then upgraded to 3.2 this is not true. Should we mention this only applies if you started your project from 3.2?

File: docs/files/integration.rst [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+take is to enable them through configuration.
+
+In case you want all the modules automatically installed you can navigate to
+your project and with your virtual environment selected you can
+
+.. code-block:: console
+
+   $ pip install invenio[auth]
```
Comment (equadon) [inline docs/files/integration.rst]: Did you mean `[all]`?

File: docs/files/integration.rst [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+`invenio-files-rest <https://invenio-files-rest.readthedocs.io/>`_
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code-block:: console
+
+   $ pip install invenio-files-rest
+
+:code:`invenio-files-rest` is the first required module for managing files
+with your application, and some of its key functions is to allow you to store
```
Comment (equadon) [inline docs/files/integration.rst]:
```suggestion
with your application, and one of its key functions is to allow you to store
```

File: docs/files/integration.rst [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+and retrieve files in a similar way to Amazon S3 APIs.
+
+- Configurable files storage
+- Secure REST APIs for upload/download
+- Support for large file uploads and multipart upload.
+- File integrity monitoring
+- Customizable access control
+
+After you install the module on your environment you just have to register
```
Comment (equadon) [inline docs/files/integration.rst]:
```suggestion
After you install the module in your environment you just have to register
```

File: docs/files/integration.rst [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+the invenio-files-rest blueprint in the list of blueprints defined in the
+:code:`entry_points` of the :code:`setup.py` of your application.
+
+.. code-block:: python
+
+    entry_points={
+        "invenio_app.api_blueprints": [
+            "invenio_files = invenio_files_rest.views:blueprint",
```
Comment (equadon) [inline docs/files/integration.rst]: Was this change needed when you were testing? In the end for my upgrade I didn't have to do this.

File: docs/files/integration.rst [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+            '<endpoint-prefix>': '<endpoint-suffix>',
+        }
+    }
+
+Read more about `configuration <https://invenio-records-files.readthedocs.io/en/latest/configuration.html>`__
+and `usage <https://invenio-records-files.readthedocs.io/en/latest/usage.html>`__
+of invenio-records-files on the module's `documentation <https://invenio-records-files.readthedocs.io/>`__.
+
+
+`invenio-previewer <https://invenio-previewer.readthedocs.io/>`_
```
Comment (equadon) [inline docs/files/integration.rst]: Seems like we have worked on the same issue here. Perhaps we could merge what we wrote in one place?

File: docs/files/integration.rst [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+configuration in the :code:`config.py` of your application, for the
+endpoints you want to use, it should be in the following form
+
+.. code-block:: python
+
+    RECORDS_FILES_REST_ENDPOINTS = {
+        '<*_REST_ENDPOINTS>': {
+            '<endpoint-prefix>': '<endpoint-suffix>',
+        }
+    }
```
Comment (equadon) [inline docs/files/integration.rst]: Perhaps we can also mention that `FILES_REST_PERMISSION_FACTORY` needs to be configured?

File: docs/files/backends.rst (hunk 5 above)
Comment (ChiaraBi) [reply docs/files/backends.rst]: There's a small error in the latest part of the link: `#invenio_files_rest.ext.FileStorage` -> `#invenio_files_rest.storage.FileStorage`

**Comment by ntarocco** [comment]
Will take all comments into account.
